File: ios/HybridMailEngine.swift
//
//  HybridMailEngine.swift
//  react-native-mail-engine
//
//  Root factory HybridObject. `connect` builds + validates a MailCore2 session
//  on a background thread and returns a HybridMailAccount.
//

import Foundation
import NitroModules

final class HybridMailEngine: HybridMailEngineSpec {
  private let queue = DispatchQueue(
    label: "com.margelo.nitro.mailengine.connect",
    qos: .userInitiated
  )

  func connect(config: MailAccountConfig) throws -> Promise<any HybridMailAccountSpec> {
    let imap = config.imap
    let smtp = config.smtp
    let auth = config.auth
    let timeoutSec = (config.connectionTimeoutMs ?? 30_000) / 1_000.0
    let accountId = config.id ?? UUID().uuidString

    return Promise.parallel(queue) { () -> (any HybridMailAccountSpec) in
      do {
        let session = try RNMailSession.connect(
          withImapHost: imap.host,
          imapPort: UInt32(imap.port),
          securityImap: securityInt(imap.security),
          allowInvalidImapTls: imap.allowInvalidCertificates ?? false,
          smtpHost: smtp?.host,
          smtpPort: UInt32(smtp?.port ?? 0),
          securitySmtp: securityInt(smtp?.security ?? .tls),
          allowInvalidSmtpTls: smtp?.allowInvalidCertificates ?? false,
          authType: authInt(auth.type),
          username: auth.user,
          password: auth.password,
          accessToken: auth.accessToken,
          connectTimeoutSec: timeoutSec
        )
        return HybridMailAccount(session: session, id: accountId)
      } catch let error as NSError {
        throw bridgeError(error)
      }
    }
  }
}

File: android/src/main/java/com/margelo/nitro/mailengine/HybridMailEngine.kt
package com.margelo.nitro.mailengine

import com.margelo.nitro.core.Promise
import java.util.UUID
import java.util.concurrent.Executors
import javax.mail.Session
import javax.mail.Store

/**
 * Root factory HybridObject. Connects + authenticates an IMAP [Store] on a fresh
 * single-thread executor (JavaMail's Store/Folder objects are NOT thread-safe,
 * so every operation for the resulting account — and its mailboxes — runs on
 * that one thread).
 */
class HybridMailEngine : HybridMailEngineSpec() {
  override fun connect(config: MailAccountConfig): Promise<HybridMailAccountSpec> {
    val executor = Executors.newSingleThreadExecutor { r ->
      Thread(r, "rn-mail-engine-${config.id ?: "account"}")
    }
    return MailBridge.run<HybridMailAccountSpec>(executor) {
      val timeout = (config.connectionTimeoutMs ?: 30_000.0).toInt()
      val (props, protocol) = MailBridge.buildProperties(
        protocol = "imap",
        host = config.imap.host,
        port = config.imap.port.toInt(),
        security = config.imap.security,
        authType = config.auth.type,
        allowInvalidCerts = config.imap.allowInvalidCertificates ?: false,
        timeoutMs = timeout
      )
      val session = Session.getInstance(props, null)
      val store: Store = session.getStore(protocol)
      try {
        store.connect(
          config.imap.host,
          config.imap.port.toInt(),
          config.auth.user,
          MailBridge.secretFor(config.auth)
        )
      } catch (error: Throwable) {
        // Graceful shutdown: we're running on this executor's own thread, so
        // shutdownNow() would interrupt ourselves. shutdown() lets this task finish.
        executor.shutdown()
        throw error
      }
      HybridMailAccount(
        store = store,
        smtp = config.smtp,
        auth = config.auth,
        accountId = config.id ?: UUID.randomUUID().toString(),
        executor = executor
      )
    }
  }
}
